[GPU] Implement per-token FC dyn-quan

byungilm opened the pull request:
### Details:
 - *item1*
 - *...*

### Tickets:
 - 158513

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -840,17 +895,19 @@ void FullyConnected_bf_tiled::GetUpdateDispatchDataFunc(KernelData& kd) const {
                     kd.kernels[0].skip_execution = false;
                     size_t input_f = get_input_bf_size(prim_params).second;
                     size_t input_size = input_f * dispatchData.tile_m * dispatchData.gws[2];
+                    size_t quan_var_size = (input_size / quantize_grp_size) * 4 * 2;
```
what about `(input_size / quantize_grp_size) * sizeof(float) * 2` for better readability?

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -840,17 +895,19 @@ void FullyConnected_bf_tiled::GetUpdateDispatchDataFunc(KernelData& kd) const {
 
-                    if (kd.internalBufferSizes[0] < input_size) {
+                    if (kd.internalBufferSizes[0] < input_size ||
+                        kd.internalBufferSizes[1] < quan_var_size || true) {
```
unnecessary `true`

isanghao commented on src/plugins/intel_gpu/tests/unit/test_cases/fully_connected_gpu_test.cpp:
```diff
@@ -2940,7 +2940,7 @@ class fully_connected_gpu_tests: public ::testing::Test {
             ASSERT_TRUE(impl != NULL);
             auto kernel_num = (is_dynamic) ? 3 : 2;
-            kernel_num = (quantize_group_size < 32) ? 2 : kernel_num;
+            kernel_num = (quantize_group_size < 32) ? ((quantize_group_size != -1) ? 2 : kernel_num) : kernel_num;
```
it would be better to change `quantize_group_size` to uint64_t. Then you can simplify this line I guess.

byungilm commented on src/plugins/intel_gpu/src/kernel_selector/cl_kernels/fully_connected_gpu_bf_tiled.cl:
```diff
@@ -992,8 +990,8 @@ inline void FUNC(fc_bf_tiled_kernel_dyn_quan)(
-                de_quantize_scale[bi * 2] = quan_var[scale_offset * 2];
-                de_quantize_scale[bi * 2 + 1] = quan_var[scale_offset * 2 + scale_pitch * 2];
+                de_quantize_scale[bi * 2] = TO_INPUT0_TYPE(quan_var[scale_offset * 2]);
+                de_quantize_scale[bi * 2 + 1] = TO_INPUT0_TYPE(quan_var[scale_offset * 2 + scale_pitch * 2]);
```
main issue is activation_sum.

byungilm commented on src/plugins/intel_gpu/src/kernel_selector/cl_kernels/fully_connected_gpu_bf_tiled.cl:
```diff
@@ -17,51 +17,52 @@
     const __global INPUT0_TYPE* input,
     __global DQ_TYPE* quantized_input,
-    __global INPUT0_TYPE* quan_var
+    __global float* quan_var
```
Yes, so I listed-up optimizing quantizing kernel. But still quantizing kernel is much smaller than FC kernel.

byungilm commented on src/plugins/intel_gpu/src/kernel_selector/cl_kernels/fully_connected_gpu_bf_tiled.cl:
```diff
@@ -17,51 +17,52 @@
     const __global INPUT0_TYPE* input,
     __global DQ_TYPE* quantized_input,
-    __global INPUT0_TYPE* quan_var
+    __global float* quan_var
```
quan_var contains activation_sum. In case of per-token, sum can not be in a range of half.

byungilm commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -87,16 +108,45 @@ static size_t get_dynamic_quantize_group_size(const fully_connected_params& para
+        zp_group_size = params.weights.IFM().v / params.decompression_zero_point.Feature().v;
+
+    // Per-token dyn-quan
+    if (dynamic_quantization_group_size >= min_quantize_grp_size && is_per_token_dynamic_quantize(params)) {
```
It is not to ignore debugging feature 'dynamic_quantize_layers_without_onednn'.
dynamic_quantization_group_size can be 0 at this line by the feature.

byungilm commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -105,7 +155,7 @@ static size_t get_dynamic_quantize_group_size(const fully_connected_params& para
 
 static bool should_dynamic_quantize(const fully_connected_params& params, bool print_log = false) {
-    size_t dynamic_quantization_group_size = get_dynamic_quantize_group_size(params);
+    size_t dynamic_quantization_group_size = get_dynamic_quantize_group_size(params, print_log);
```
Applied

byungilm commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -709,10 +762,12 @@ JitConstants FullyConnected_bf_tiled::GetJitConstants(const fully_connected_para
     }
-    jit.AddConstant(MakeJitConstant("DQ_TYPE", "char"));
 
+    jit.AddConstant(MakeJitConstant("INPUT_LOAD_SIZE", act_load_size));
```
Applied

byungilm commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -87,16 +108,45 @@ static size_t get_dynamic_quantize_group_size(const fully_connected_params& para
+            // For int8 ASYM model, activation_sum should fit to weight zp
+            if (is_dyn_quan_8bit_asym(params) && params.has_decompression_zp == true &&
+                dynamic_quantization_group_size > zp_group_size && (zp_group_size % act_load_size) == 0) {
+                dynamic_quantization_group_size = zp_group_size;
```
This is just exception for possibility. I added it looked difficult to find later if it happens.
Does It have no possibility that zp does not match to scale_goup_size?

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -87,16 +108,45 @@ static size_t get_dynamic_quantize_group_size(const fully_connected_params& para
+
+            GPU_DEBUG_TRACE_DETAIL << "FC dyn-quantize by per-token. Actual dyn_quan_group_size(" << dynamic_quantization_group_size
+                                    << ") : From scale_group_size (" << scale_group_size << ", zp_group_size("  << zp_group_size
+                                    << "), zp_group_num(" << zp_group_num << "), ifm_size (" << get_input_bf_size(params).second << ")" << std::endl;
```
* Please use `LOG` instead of `TRACE_DETAIL`

byungilm replied:
Applied

byungilm commented:
Resolved some conflicts from fc_gpu_bf_tile.cpp

byungilm commented:
+ Reverted change of data type 'quan_var'. (It assums that activation sum of sym 8bit input would not overflow from half range.
+ Changed important log to _LOG level about setting group size.
+

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/cl_kernels/fully_connected_gpu_bf_tiled.cl:
```diff
@@ -28,38 +26,39 @@ KERNEL(quantize_input)(
 
     // Pair of quantizing_scale and quantized activation_sum for each group
-    quan_var[offset * 2] = quan_scale;
+    quan_var[offset * 2] = CAT(CAT(convert_, INPUT0_TYPE), _rte)(quan_scale);
```
I guess we don't need `_rte` for `convert_half` function

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/cl_kernels/fully_connected_gpu_bf_tiled.cl:
```diff
@@ -991,25 +1014,25 @@ inline void FUNC(fc_bf_tiled_kernel_dyn_quan)(
-                    activation_sum[bi * 2] = quan_var[scale_offset * 2 + 1];
-                    activation_sum[bi * 2 + 1] = quan_var[scale_offset * 2 + 1 + scale_pitch * 2];
+                    activation_sum[bi * 2] = TO_INPUT0_TYPE(quan_var[scale_offset * 2 + 1]);
+                    activation_sum[bi * 2 + 1] = TO_INPUT0_TYPE(quan_var[scale_offset * 2 + 1 + scale_pitch * 2]);
```
I think you reverted the change to use `float` for `quan_var`. Then I guess we don't need `TO_INPUT0_TYPE` here.

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/cl_kernels/fully_connected_gpu_bf_tiled.cl:
```diff
@@ -1208,7 +1217,7 @@ inline void FUNC(fc_bf_tiled_kernel_dyn_quan)(
         }  // Whole tile_k elements of each iteration : ki
 
-        #if DQ_DECOMPRESSION_SCALE_POST_OP && (TILE_IFM_ELEMENTS_SIZE <= DECOMPRESSION_SCALE_GROUP_SIZE)
+        #if (PER_TOKEN_SIZE_DYN_QUANTIZE == 0) && DQ_DECOMPRESSION_SCALE_POST_OP && (TILE_IFM_ELEMENTS_SIZE <= DECOMPRESSION_SCALE_GROUP_SIZE)
```
nit: what about `!PER_TOKEN_SIZE_DYN_QUANTIZE`

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -52,17 +53,37 @@ static std::pair<size_t, size_t> get_output_aligned_bf_size(const fully_connecte
+    return params.weights.IFM().v / params.decompression_scale.Feature().v;
+}
+
+static bool is_dyn_quan_8bit_asym(const fully_connected_params& params) {
```
* we need to check `asym` with zp, not the data type. JFYI, Sergey implemented asymmetric quantization with i8.
* What about using more clear name like `8bit_asym_wei`?

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -52,17 +53,37 @@ static std::pair<size_t, size_t> get_output_aligned_bf_size(const fully_connecte
         return true;
-    // UINT8 weight type is supported by FC dyn-quantize(with SLM).
-    if (weight_type == WeightsType::UINT8)
+    if (is_dyn_quan_8bit_asym(params))
```
why symmetric 8bit weight is not supported? is it because you did not see such case? then could you please add some comment here?

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -703,16 +752,23 @@ JitConstants FullyConnected_bf_tiled::GetJitConstants(const fully_connected_para
+        if (is_per_token_dynamic_quantize(params) && quantize_grp_size == get_input_bf_size(params).second)
+            jit.AddConstant(MakeJitConstant("PER_TOKEN_SIZE_DYN_QUANTIZE", 1));
+        else
+            jit.AddConstant(MakeJitConstant("PER_TOKEN_SIZE_DYN_QUANTIZE", 0));
```
nit: what about 
`jit.AddConstant(MakeJitConstant("PER_TOKEN_SIZE_DYN_QUANTIZE", is_per_token_dynamic_quantize(params) && quantize_grp_size == get_input_bf_size(params).second));`

isanghao commented on src/plugins/intel_gpu/src/kernel_selector/kernels/fully_connected/fully_connected_kernel_bf_tiled.cpp:
```diff
@@ -840,19 +896,20 @@ void FullyConnected_bf_tiled::GetUpdateDispatchDataFunc(KernelData& kd) const {
                     size_t input_f = get_input_bf_size(prim_params).second;
                     size_t input_size = input_f * dispatchData.tile_m * dispatchData.gws[2];
+                    OPENVINO_ASSERT(quantize_grp_size != 0, "Error: quantize_grp_size is zero.");
+                    size_t quan_var_size = (input_size / quantize_grp_size) * sizeof(float) * 2;
```
I guess it is now not `float`, but `half`.
